backends/text-to-speech/linux/linux-text-to-speech.cpp

```diff
@@ -0,0 +1,393 @@
+		}
+	}
+
+	return false;
+}
+
+bool LinuxTextToSpeechManager::stop() {
+	if (_speechState == READY || _speechState == BROKEN)
+		return true;
+	_speechState = READY;
+	pthread_mutex_lock(&_speechMutex);
+	_speechQueue.clear();
+	bool result = spd_cancel(_connection) == -1;
+	pthread_mutex_unlock(&_speechMutex);
+	return result;
+}
+
+bool LinuxTextToSpeechManager::pause() {
+	if (_speechState == READY || _speechState == PAUSED || _speechState == BROKEN)
+		return true;
+	pthread_mutex_lock(&_speechMutex);
+	_speechState = PAUSED;
+	bool result = spd_cancel_all(_connection) == -1;
+	pthread_mutex_unlock(&_speechMutex);
+	if (result)
+		return true;
+	return false;
+}
+
+bool LinuxTextToSpeechManager::resume() {
+	if (_speechState == READY || _speechState == SPEAKING || _speechState == BROKEN)
+		return true;
+	// If there is a thread from before pause() waiting, let it finish (it shouln't
+	// do anything). There shouldn't be any other threads getting created,
+	// because the speech is paused, so we don't need to synchronize
+	if (_threadCreated) {
+		pthread_join(_thread, NULL);
+		_threadCreated = false;
+	}
+	_speechState = PAUSED;
+	if (_speechQueue.size()) {
+		_speechState = SPEAKING;
+		startSpeech((void *) &_params);
+	}
+	else
+		_speechState = READY;
+	return false;
+}
+
+bool LinuxTextToSpeechManager::isSpeaking() {
+	return _speechState == SPEAKING;
+}
+
+bool LinuxTextToSpeechManager::isPaused() {
+	return _speechState == PAUSED;
+}
+
+bool LinuxTextToSpeechManager::isReady() {
+	return _speechState == READY;
+}
+
+void LinuxTextToSpeechManager::setVoice(unsigned index) {
+	if (_speechState == BROKEN)
+		return;
+	assert(index < _ttsState->_availableVoices.size());
+	Common::TTSVoice voice = _ttsState->_availableVoices[index];
+	spd_set_voice_type(_connection, *(SPDVoiceType *)(voice.getData()));
+	_ttsState->_activeVoice = index;
+}
+
+void LinuxTextToSpeechManager::setRate(int rate) {
+	if (_speechState == BROKEN)
+		return;
+	assert(rate >= -100 && rate <= 100);
+	spd_set_voice_rate(_connection, rate);
+	_ttsState->_rate = rate;
+}
+
+void LinuxTextToSpeechManager::setPitch(int pitch) {
+	if (_speechState == BROKEN)
+		return;
+	assert(pitch >= -100 && pitch <= 100);
+	spd_set_voice_pitch(_connection, pitch);
+	_ttsState->_pitch = pitch;
+}
+
+void LinuxTextToSpeechManager::setVolume(unsigned volume) {
+	if (_speechState == BROKEN)
+		return;
+	assert(volume <= 100);
+	spd_set_volume(_connection, (volume - 50) * 2);
+	_ttsState->_volume = volume;
+}
+
+void LinuxTextToSpeechManager::setLanguage(Common::String language) {
+	if (_speechState == BROKEN)
+		return;
+	spd_set_language(_connection, language.c_str());
+	_ttsState->_language = language;
+	setVoice(_ttsState->_activeVoice);
+}
+
+void LinuxTextToSpeechManager::createVoice(int typeNumber, Common::TTSVoice::Gender gender, Common::TTSVoice::Age age, char *description) {
+	SPDVoiceType *type = (SPDVoiceType *) malloc(sizeof(SPDVoiceType));
```

Comment:
Because I need to pass the vaule to the TTSVoice class as it's data field, which is defined as void *

The data field in the TTSVoice is meant for platform specific information about each voice. The WindowsTextToSpeechManager for example saves there a pointer to a whole class ISpObjectToken.

So because I needed a pointer to the value and I can't just pass a pointer to the stack, because the TTSVoice that is being created in this function outlives this function, I alocate space for it on the heap and copy it there.

Reply:
Thanks for the explanation. Could you please add a corresponding comment to the code?